web/source/kmwosk.js

```diff
@@ -2592,36 +2974,73 @@ if(!window['tavultesoft']['keymanweb']['initialized']) {
+            // Only create keys for defined layers.  ('default' and 'shift' are always defined.)
+            if(layerSpec || isDefault) {
+              // Get keycap text from visual keyboard array, if defined in keyboard
+              if(layerSpec) {
+                if(kx >= 0 && kx < layerSpec.length) key['text']=layerSpec[kx];
+              }
 
-            // Fall back to US English keycap text (any 'ghost' keys must now be explicitly defined in layout)
-            if(key['text'] == '' &&  key['id'] != 'K_SPACE' && kx+65*n < dfltText.length) key['text']=dfltText[kx+65*n];
+              // Fall back to US English keycap text as default for the base two layers if not otherwise defined.
+              // (Any 'ghost' keys must be explicitly defined in layout for these layers.)
+              if(isDefault) {
+                if((key['text'] == '' || typeof key['text'] == 'undefined') &&  key['id'] != 'K_SPACE' && kx+65 * isShift < dfltText.length) {
+                  key['text']=dfltText[kx+65*isShift];
+                }
+              }
+            }
 
             // Leave any unmarked key caps as null strings
-            if(typeof(key['text']) == 'undefined') key['text']='';
+            if(typeof(key['text']) == 'undefined') { 
+              key['text']='';
+            }
+            // Detect important tracking keys.
+            switch(key['id']) {
+              case "K_SHIFT":
+                shiftKey=key;
+                break;
+              case "K_TAB":
+                nextKey=key;
+                break;
+              case "K_CAPS":
+                capsKey=key;
+                break;
+              case "K_NUMLOCK":
+                numKey=key;
+                break;
+              case "K_SCROLL":
+                scrollKey=key;
+                break;
+            }
 
-            if(key['id'] == 'K_SHIFT') shiftKey=key;
-            if(key['id'] == 'K_TAB') nextKey=key;
+            // Remove pop-up shift keys referencing invalid layers (Build 349)
+            if(key['sk'] != null) {
+              for(k=0; k<key['sk'].length; k++) {
+                if(validIdList.indexOf(key['sk'][k]['nextlayer']) == -1) {
+                  key['sk'].splice(k--, 1);
+                }
+              }
 
-            // Concatenate all keys for this layer, excluding special keys, to check for existence of layer
-            if(!specialLabel.test(key['text'])) allText=allText+key['text'];
+              if(key['sk'].length == 0) {
+                key['sk']=null;
+              }
+            }
           }
         }
 
-        // Mark layer as valid if any key caps (excluding special keys) are defined, invalid otherwise
-        layer.valid=(allText.trim().length > 0);
+        // We're done with the layer keys initialization pass.  Time to do post-analysis layer-level init where necessary.
         layer.shiftKey=shiftKey;
+        layer.capsKey=capsKey;
```

Comment:
Why are we adding Caps, Num and Scroll Keys? What do they do given they can't affect physical key state?

Reply:
The reason for the properties is to facilitate their OSK feedback.  The base modifiers are selectively highlighted based upon the layer on which they appear, so they never needed to be tracked for any special purpose.  The state keys, however, don't have their own layer.  If we want to say a given state key is toggled, we need a way to forcibly toggle it for whatever layer is presently displayed.  See line 1744 for their use in this matter.

Note that toggling a standard modifier within the OSK similarly cannot affect physical key state - it never has within KeymanWeb.  The only difference is that the modifier keys have corresponding layers.

Reply:
I'm not quite sure I'm following you. Are you saying that these state keys on the OSK can be toggled by the user, or are they purely for visual feedback?

Side note: the difficulty with state keys is that the user will expect their, um, state, to persist across keyboard changes. So if they press Caps Lock on the OSK, they won't anticipate Caps Lock being off once they click in the Address Bar, for example.

Reply:
They can be "toggled" (simulated) for OSK input.  This does not affect the true key state whatsoever, and any physical keyboard input will instantly reset the state.  This reflects how the modifiers are presently handled.